server/src/main/java/org/elasticsearch/index/mapper/SourceFieldMapper.java

```diff
@@ -59,8 +60,17 @@ public class SourceFieldMapper extends MetadataFieldMapper {
 
     public static final String LOSSY_PARAMETERS_ALLOWED_SETTING_NAME = "index.lossy.source-mapping-parameters";
 
+    public static final Setting<Mode> INDEX_MAPPER_SOURCE_MODE_SETTING = Setting.enumSetting(
+        SourceFieldMapper.Mode.class,
+        "index.mapping.source.mode",
+        Mode.UNDEFINED,
+        Setting.Property.Final,
+        Setting.Property.IndexScope
+    );
+
     /** The source mode */
-    private enum Mode {
+    public enum Mode {
+        UNDEFINED,
```

Comment:
We will get rid of this enum value once we deprecate the mapping level `_source.mode`. In that case we can decide its value based just on `index.mode` and `index.mapper.source.mode`. In the meanwhile, until we keep the `_source.mode` it needs to take precedence if the setting is not defined.